src/chords.py

```diff
@@ -0,0 +1,164 @@
+import queue
+
+import numpy as np
+import pandas as pd
+import fortepyan as ff
+from tqdm import tqdm
+import chorder.dechorder
+import chorder.timepoints
+from datasets import load_dataset
+from miditoolkit.midi.containers import Note
+from pretty_midi.pretty_midi import PrettyMIDI
+
+from helpers import plot
+
+
+def start_to_ticks(row, piece: PrettyMIDI):
+    return piece.time_to_tick(row["start"])
+
+
+def end_to_ticks(row, piece: PrettyMIDI):
+    return piece.time_to_tick(row["end"])
+
+
+def end_to_time(row, piece: PrettyMIDI):
+    return piece.tick_to_time(row["end"])
+
+
+def get_chords(piece: ff.MidiPiece):
+    """
+    Extracts chords from a MidiPiece object.
+    Treats every group of simultaneously playing notes,
+    which can be interpreted as a chord by a dechorder, as a chord.
+
+    Parameters:
+        piece (MidiPiece): A MidiPiece object containing musical events.
+
+    Returns:
+        list: A list of detected chords in the format [chord_name, start_time, end_time].
+    """
+    # mido_obj = midipiece_to_midifile(piece)
+    # define comparator
+    setattr(Note, "__lt__", lambda self, other: self.end <= other.end)
+    pq = queue.PriorityQueue()
+    start = 0
+    count = 0
+    chords = []
+    prev_chord = None
+    midi_data = piece.df
+    piece = piece.to_midi()
+    midi_data["start"] = midi_data.apply(start_to_ticks, axis=1, args=(piece,))
+    midi_data["end"] = midi_data.apply(end_to_ticks, axis=1, args=(piece,))
```

Comment:
```suggestion
    midi_data = piece.df.copy()
    piece = piece.to_midi()
    midi_data["start"] = midi_data.apply(start_to_ticks, axis=1, args=(piece,))
    midi_data["end"] = midi_data.apply(end_to_ticks, axis=1, args=(piece,))
```

Your version modifies the internal dataframe of the piece, which could lead to serious confusion if the same object had to be used for any other processing after chord detection. Making a copy is a safe way to prevent those problems.

Reply:
Thank you very much!